qiskit/backends/_projectq_simulator.py

```diff
@@ -0,0 +1,256 @@
+                              QubitOperator,
+                              TimeEvolution,
+                              All)
+logger = logging.getLogger(__name__)
+
+
+class ProjectQSimulator(BaseBackend):
+    """Python interface to Project Q simulator"""
+
+    def __init__(self, configuration=None):
+        """
+        Args:
+            configuration (dict): backend configuration
+        """
+        if CppSim is None:
+            logger.info('Project Q C++ simulator unavailable.')
+            raise ImportError('Project Q C++ simulator unavailable.')
+        if configuration is None:
+            self._configuration = {
+                'name': 'local_projectq_simulator',
+                'url': 'https://projectq.ch',
+                'simulator': True,
+                'local': True,
+                'description': 'ProjectQ C++ simulator',
+                'coupling_map': 'all-to-all',
+                'basis_gates': 'h,s,t,cx,id'
+            }
+        else:
+            self._configuration = configuration
+
+    def run(self, q_job):
+        """Run circuits in q_job"""
+        # Generating a string id for the job
+        job_id = str(uuid.uuid4())
+        result_list = []
+        qobj = q_job.qobj
+        self._sim = Simulator(gate_fusion=True)
+        if 'seed' in qobj['config']:
+            self._seed = qobj['config']['seed']
+            self._sim._simulator = CppSim(self._seed)
+        else:
+            self._seed = random.getrandbits(32)
+        self._shots = qobj['config']['shots']
+        for circuit in qobj['circuits']:
+            result_list.append(self.run_circuit(circuit))
+        return Result({'job_id': job_id, 'result': result_list,
+                       'status': 'COMPLETED'},
+                      qobj)
+
+    def run_circuit(self, circuit):
+        """Run a circuit and return a single Result.
+
+        Args:
+            circuit (dict): JSON circuit from qobj circuits list
+
+        Returns:
+            A dictionary of results which looks something like::
+
+                {
+                "data":
+                    {  #### DATA CAN BE A DIFFERENT DICTIONARY FOR EACH BACKEND ####
+                    "counts": {’00000’: XXXX, ’00001’: XXXXX},
+                    "time"  : xx.xxxxxxxx
+                    },
+                "status": --status (string)--
+                }
+        """
+        ccircuit = circuit['compiled_circuit']
+        self._number_of_qubits = ccircuit['header']['number_of_qubits']
+        self._number_of_clbits = ccircuit['header']['number_of_clbits']
+        self._quantum_state = 0
+        self._classical_state = 0
+        cl_reg_index = [] # starting bit index of classical register
+        cl_reg_nbits = [] # number of bits in classical register
+        clbit_index = 0
+        qobj_quregs = OrderedDict(_get_register_specs(
+            ccircuit['header']['qubit_labels']))
+        eng = MainEngine(backend=self._sim)
```

Comment:
Changing the MainEngine to
```
eng = MainEngine(backend=self._sim, engine_list=[])
```
This would turn off the compiler engines, i.e. send all gates directly to the backend which involves a lot less overhead and hence a small change to get more performance out of the simulator.

Reason:
If engine_list is not specified the default are engines which break gates with more than 5 qubits down to into smaller gates (simulator can natively perform up to 5 qubits gates with an arb. amount of control qubits) and engines to optimise the circuit. Both seem to be an overkill when just using the ProjectQ simulator in this way...

Another option:
Completely remove the ProjectQ front-end "MainEngine" which would make the code even faster as @thomashaener has previously mentioned.

The changes for this would be minimal (just follow https://github.com/ProjectQ-Framework/ProjectQ/blob/develop/projectq/backends/_sim/_simulator.py):
1) instead of getting qubits via eng.allocate_qureg(), just use integers 0,1,2,... as IDs for qubits and call the simulator directly to allocate these qubits, e.g.,
```
for qubit_id in range(n):
    self._sim.allocate_qubit(qubit_id)
```
2) Apply the gates directly to the simulator which means changing:
```
Rz(params[2]) | qubit
```
to 
```
self._sim.apply_controlled_gate(Rz(params[2]).tolist(), qubit_id, [])
```
or for a CNOT:
```
self._sim.apply_controlled_gate(X.tolist(), qubit_id, [control_qubit_id])
```
3.  Apply measurements directly:
```
measurement_result = self._sim.measure_qubits(ids) #ids is a list of integers
```
4. instead of `eng.flush()` call `self._sim.run()`

This would save a few if branches and one function call from eng to self._sim which the currently implemented version has. Please let us (@thomashaener) know if you need help should you decide to go this route...